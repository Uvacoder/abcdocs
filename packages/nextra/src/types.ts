import { NextConfig } from 'next'
import { Heading as MDASTHeading } from 'mdast'
import { ProcessorOptions } from '@mdx-js/mdx'
import { Options as RehypePrettyCodeOptions } from 'rehype-pretty-code'
import { GrayMatterFile } from 'gray-matter'

export abstract class NextraPluginCache {
  public cache: { items: PageMapItem[]; fileMap: Record<string, any> } | null

  constructor() {
    this.cache = { items: [], fileMap: {} }
  }

  set(data: { items: PageMapItem[]; fileMap: Record<string, any> }) {
    this.cache!.items = data.items
    this.cache!.fileMap = data.fileMap
  }

  clear() {
    this.cache = null
  }

  get() {
    return this.cache
  }
}
export interface LoaderOptions extends NextraConfig {
  pageImport?: boolean
  locales: string[]
  defaultLocale: string
  pageMapCache: NextraPluginCache
  newNextLinkBehavior?: boolean
}

export interface PageMapItem {
  name: string
  route: string
  locale?: string
  children?: PageMapItem[]
  timestamp?: number
  frontMatter?: Record<string, any>
  meta?: Record<string, any>
  active?: boolean
}

export type Heading = MDASTHeading & {
  value: string
}

export type PageOpts = {
  filePath: string
  route: string
  frontMatter: GrayMatterFile<string>['data']
  pageMap: PageMapItem[]
  title: string
  headings: Heading[]
  hasJsxInH1?: boolean
  timestamp?: number
  unstable_flexsearch?: Flexsearch
  newNextLinkBehavior?: boolean
}

export type PageMapResult = [
  pageMap: PageMapItem[],
  route: string,
  title: string
]

type Theme = string
type Flexsearch = boolean | { codeblocks: boolean }

export type NextraConfig = {
  theme: Theme
  themeConfig?: string
  unstable_flexsearch?: Flexsearch
  unstable_staticImage?: boolean
  mdxOptions?: Pick<ProcessorOptions, 'rehypePlugins' | 'remarkPlugins'> & {
    rehypePrettyCodeOptions?: Partial<RehypePrettyCodeOptions>
  }
}

export type Nextra = (
  ...args: [NextraConfig] | [theme: Theme, themeConfig: string]
) => (nextConfig: NextConfig) => NextConfig

const nextra: Nextra =
  (...args) =>
  nextConfig => ({})

export default nextra
