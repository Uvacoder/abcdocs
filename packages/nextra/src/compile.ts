import { createProcessor, ProcessorOptions } from '@mdx-js/mdx'
import { Processor } from '@mdx-js/mdx/lib/core'
import remarkGfm from 'remark-gfm'
import rehypePrettyCode from 'rehype-pretty-code'
import { rehypeMdxTitle } from 'rehype-mdx-title'
import { remarkStaticImage } from './mdx-plugins/static-image'
import { remarkHeadings } from './mdx-plugins/remark'
import { LoaderOptions, PageOpts } from './types'
import structurize from './mdx-plugins/structurize'
import { parseMeta, attachMeta } from './mdx-plugins/rehype-handler'
import theme from './theme.json'
import { truthy } from './utils'

const createCompiler = (mdxOptions: ProcessorOptions): Processor => {
  const compiler = createProcessor(mdxOptions)
  compiler.data('headingMeta', {
    headings: []
  })
  return compiler
}

const rehypePrettyCodeOptions = {
  theme,
  onVisitLine(node: any) {
    // Prevent lines from collapsing in `display: grid` mode, and
    // allow empty lines to be copy/pasted
    if (node.children.length === 0) {
      node.children = [{ type: 'text', value: ' ' }]
    }
  },
  onVisitHighlightedLine(node: any) {
    node.properties.className.push('highlighted')
  },
  onVisitHighlightedWord(node: any) {
    node.properties.className = ['highlighted']
  }
}

export async function compileMdx(
  source: string,
  mdxOptions: LoaderOptions['mdxOptions'] &
    Pick<ProcessorOptions, 'jsx' | 'outputFormat'> = {},
  nextraOptions: Pick<
    LoaderOptions,
    | 'unstable_staticImage'
    | 'unstable_flexsearch'
    | 'unstable_defaultShowCopyCode'
  > = {},
  filePath = ''
) {
  const structurizedData = Object.create(null)
  const compiler = createCompiler({
    jsx: mdxOptions.jsx ?? true,
    outputFormat: mdxOptions.outputFormat,
    providerImportSource: '@mdx-js/react',
    remarkPlugins: [
      ...(mdxOptions.remarkPlugins || []),
      remarkGfm,
      remarkHeadings,
      nextraOptions.unstable_staticImage && remarkStaticImage,
      nextraOptions.unstable_flexsearch &&
        structurize(structurizedData, nextraOptions.unstable_flexsearch)
    ].filter(truthy),
    rehypePlugins: [
      ...(mdxOptions.rehypePlugins || []),
      [
        parseMeta,
        { defaultShowCopyCode: nextraOptions.unstable_defaultShowCopyCode }
      ],
      [
        rehypePrettyCode,
        { ...rehypePrettyCodeOptions, ...mdxOptions.rehypePrettyCodeOptions }
      ],
      [rehypeMdxTitle, { name: '__nextra_title__' }],
      attachMeta
    ]
  })
  try {
    const result = String(await compiler.process(source))
      .replace('export const __nextra_title__', 'const __nextra_title__')
      .replace('export default MDXContent;', '')

    return {
      result,
      ...(compiler.data('headingMeta') as Pick<
        PageOpts,
        'headings' | 'hasJsxInH1'
      >),
      structurizedData
    }
  } catch (err) {
    console.error(`[nextra] Error compiling ${filePath}.`)
    throw err
  }
}
