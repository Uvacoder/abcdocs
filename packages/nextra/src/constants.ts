export const MARKDOWN_EXTENSION_REGEX = /\.mdx?$/

export const IS_PRODUCTION = process.env.NODE_ENV === 'production'

export const LOCALE_REGEX = /\.([a-z]{2}(-[A-Z]{2})?)$/

export const DEFAULT_LOCALE = 'en-US'

export const OFFICIAL_THEMES = [
  'nextra-theme-docs',
  'nextra-theme-blog'
] as const

export const META_FILENAME = '_meta.json'

export const CWD = process.cwd()

export const MARKDOWN_EXTENSIONS = ['md', 'mdx'] as const
