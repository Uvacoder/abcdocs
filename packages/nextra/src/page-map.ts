import { PageMapItem } from './types'
import { parseFileName } from './utils'
import path from 'path'
import filterRouteLocale from './filter-route-locale'
import { META_FILENAME } from './constants'

export function getPageMap(
  currentResourcePath: string,
  pageMap: PageMapItem[],
  fileMap: Record<string, PageMapItem>,
  defaultLocale: string
): [PageMapItem[], string, string] {
  const { locale } = parseFileName(currentResourcePath)
  const pageItem = fileMap[currentResourcePath]
  const [metaName, metaExt] = META_FILENAME.split('.')
  const metaPath = path.dirname(currentResourcePath)
  const metaFilename = `${metaName}.${locale && `${locale}.`}${metaExt}`
  const pageMeta = fileMap[`${metaPath}/${metaFilename}`]?.meta?.[pageItem.name]
  const title =
    (typeof pageMeta === 'string' ? pageMeta : pageMeta?.title) || pageItem.name

  return [
    locale ? filterRouteLocale(pageMap, locale, defaultLocale) : pageMap,
    pageItem.route,
    title
  ]
}
