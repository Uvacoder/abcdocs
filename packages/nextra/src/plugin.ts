import { NextraConfig, PageMapItem, NextraPluginCache } from './types'
import fs from 'graceful-fs'
import { promisify } from 'util'
import { parseFileName, parseJsonFile } from './utils'
import path from 'path'
import slash from 'slash'
import grayMatter from 'gray-matter'
import { findPagesDir } from 'next/dist/lib/find-pages-dir.js'
import { Compiler } from 'webpack'

import { restoreCache } from './content-dump'
import { MARKDOWN_EXTENSION_REGEX, META_FILENAME } from './constants'

const readdir = promisify(fs.readdir)
const readFile = promisify(fs.readFile)

export const collectMdx = async (filePath: string, route = '') => {
  const { name, locale } = parseFileName(filePath)

  const content = await readFile(filePath, 'utf8')
  const { data } = grayMatter(content)
  return {
    name,
    route,
    locale,
    ...(Object.keys(data).length && { frontMatter: data })
  }
}

export async function collectFiles(
  dir: string,
  route = '/',
  fileMap: Record<string, any> = {}
): Promise<{ items: PageMapItem[]; fileMap: Record<string, any> }> {
  const files = await readdir(dir, { withFileTypes: true })

  const items = (
    await Promise.all(
      files.map(async f => {
        const filePath = path.resolve(dir, f.name)
        const { name, locale, ext } = parseFileName(filePath)
        const fileRoute = slash(path.join(route, name.replace(/^index$/, '')))

        if (f.isDirectory()) {
          if (fileRoute === '/api') return
          const { items } = await collectFiles(filePath, fileRoute, fileMap)
          if (!items.length) return
          return {
            name: f.name,
            children: items,
            route: fileRoute
          }
        }

        if (MARKDOWN_EXTENSION_REGEX.test(ext)) {
          fileMap[filePath] = await collectMdx(filePath, fileRoute)
          return fileMap[filePath]
        }
        const fileName = name + ext

        if (fileName === META_FILENAME) {
          const content = await readFile(filePath, 'utf8')
          fileMap[filePath] = {
            name: META_FILENAME,
            locale,
            meta: parseJsonFile(content, filePath)
          }
          return fileMap[filePath]
        }

        if (fileName === 'meta.json') {
          console.warn(
            `[nextra] "meta.json" was renamed to "_meta.json". Rename the following file:`,
            path.relative(process.cwd(), filePath)
          )
        }
      })
    )
  ).filter(Boolean)

  return {
    items,
    fileMap
  }
}

class PageMapCache implements NextraPluginCache {
  public cache: { items: PageMapItem[]; fileMap: Record<string, any> } | null

  constructor() {
    this.cache = { items: [], fileMap: {} }
  }

  set(data: { items: PageMapItem[]; fileMap: Record<string, any> }) {
    this.cache!.items = data.items
    this.cache!.fileMap = data.fileMap
  }

  clear() {
    this.cache = null
  }

  get() {
    return this.cache
  }
}

export const pageMapCache = new PageMapCache()

export class NextraPlugin {
  constructor(private config: NextraConfig) {}

  apply(compiler: Compiler) {
    compiler.hooks.beforeCompile.tapAsync(
      'NextraPlugin',
      async (_, callback) => {
        if (this.config?.unstable_flexsearch) {
          // Restore the search data from the cache.
          restoreCache()
        }
        const result = await collectFiles(
          findPagesDir(process.cwd()).pages,
          '/'
        )
        pageMapCache.set(result)
        callback()
      }
    )
  }
}
