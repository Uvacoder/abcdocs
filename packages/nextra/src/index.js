import { NextraPlugin, pageMapCache } from './plugin'
import {
  DEFAULT_LOCALE,
  MARKDOWN_EXTENSION_REGEX,
  MARKDOWN_EXTENSIONS
} from './constants'

const DEFAULT_EXTENSIONS = ['js', 'jsx', 'ts', 'tsx']

const nextra = (...config) =>
  function withNextra(nextConfig = {}) {
    const nextraConfig =
      typeof config[0] === 'string'
        ? {
            theme: config[0],
            themeConfig: config[1]
          }
        : config[0]

    const nextraPlugin = new NextraPlugin(nextraConfig)

    if (nextConfig.i18n?.locales) {
      console.log(
        '[nextra] You have Next.js i18n enabled, read here https://nextjs.org/docs/advanced-features/i18n-routing for the docs.'
      )
    }

    return {
      ...nextConfig,
      pageExtensions: [
        ...(nextConfig.pageExtensions || DEFAULT_EXTENSIONS),
        ...MARKDOWN_EXTENSIONS
      ],
      webpack(config, options) {
        config.plugins ||= []
        config.plugins.push(nextraPlugin)

        const nextraLoaderOptions = {
          ...nextraConfig,
          locales: nextConfig.i18n?.locales || [DEFAULT_LOCALE],
          defaultLocale: nextConfig.i18n?.defaultLocale || DEFAULT_LOCALE,
          pageMapCache,
          newNextLinkBehavior: nextConfig.experimental?.newNextLinkBehavior
        }

        config.module.rules.push(
          {
            // Match Markdown imports from non-pages. These imports have an
            // issuer, which can be anything as long as it's not empty.
            test: MARKDOWN_EXTENSION_REGEX,
            issuer: request => !!request,
            use: [
              options.defaultLoaders.babel,
              {
                loader: 'nextra/loader',
                options: nextraLoaderOptions
              }
            ]
          },
          {
            // Match pages (imports without an issuer).
            test: MARKDOWN_EXTENSION_REGEX,
            issuer: request => !request,
            use: [
              options.defaultLoaders.babel,
              {
                loader: 'nextra/loader',
                options: {
                  ...nextraLoaderOptions,
                  pageImport: true
                }
              }
            ]
          }
        )

        if (typeof nextConfig.webpack === 'function') {
          return nextConfig.webpack(config, options)
        }

        return config
      }
    }
  }

module.exports = nextra
