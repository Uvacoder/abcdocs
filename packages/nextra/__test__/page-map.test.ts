import { describe, it, expect, beforeAll } from 'vitest'
import path from 'node:path'
import { getPageMap } from '../src/page-map'
import { collectFiles } from '../src/plugin'
import { FileMap, PageMapItem } from '../src/types'
import { CWD } from '../src/constants'

const PAGES_DIR = path.join(CWD, '../../examples/swr-site/pages')
const filePath = (name: string) => path.join(PAGES_DIR, name)

const defaultLocale = 'en-US'

describe('Page Process', () => {
  let pageMap: PageMapItem[]
  let fileMap: FileMap
  beforeAll(async () => {
    const { items, fileMap: data } = await collectFiles(PAGES_DIR)
    pageMap = items
    fileMap = data
  })

  it('pageMap en-US', async () => {
    const indexData = await getPageMap({
      filePath: filePath('docs/data-fetching.en-US.mdx'),
      pageMap,
      fileMap,
      defaultLocale
    })
    expect([indexData.pageMap, indexData.route, indexData.title]).toMatchSnapshot()

    const gettingStartData = await getPageMap({
      filePath: filePath('docs/getting-started.en-US.mdx'),
      pageMap,
      fileMap,
      defaultLocale
    })
    expect(gettingStartData.pageMap).toEqual(indexData.pageMap)
  })

  it('pageMap zh-CN', async () => {
    const indexData = await getPageMap({
      filePath: filePath('docs/data-fetching.zh-CN.mdx'),
      pageMap,
      fileMap,
      defaultLocale
    })
    expect([indexData.pageMap, indexData.route, indexData.title]).toMatchSnapshot()

    const gettingStartData = await getPageMap({
      filePath: filePath('docs/getting-started.zh-CN.mdx'),
      pageMap,
      fileMap,
      defaultLocale
    })
    expect(gettingStartData.pageMap).toEqual(indexData.pageMap)
  })
})
