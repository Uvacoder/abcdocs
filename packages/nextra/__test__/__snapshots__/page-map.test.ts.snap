// Vitest Snapshot v1

exports[`Page Process > pageMap en-US 1`] = `
[
  [
    {
      "locale": "",
      "name": "404",
      "route": "/404",
    },
    {
      "locale": "",
      "name": "500",
      "route": "/500",
    },
    {
      "locale": "en-US",
      "meta": {
        "404": {
          "theme": {
            "timestamp": false,
          },
          "type": "page",
        },
        "500": {
          "theme": {
            "layout": "full",
            "timestamp": false,
          },
          "type": "page",
        },
        "about": {
          "items": {
            "a-page": {
              "title": "A Page",
            },
            "acknowledgement": {
              "title": "Acknowledgement",
            },
            "contributors": {
              "href": "https://github.com/vercel/swr/graphs/contributors",
              "newWindow": true,
              "title": "Contributors",
            },
            "team": {
              "title": "Team",
            },
          },
          "title": "About",
          "type": "menu",
        },
        "blog": {
          "theme": {
            "breadcrumb": false,
            "footer": true,
            "sidebar": false,
            "toc": true,
            "typesetting": "article",
          },
          "title": "Blog",
          "type": "page",
        },
        "docs": {
          "theme": {
            "toc": true,
          },
          "title": "Docs",
          "type": "page",
        },
        "examples": {
          "theme": {
            "layout": "full",
          },
          "title": "Examples",
          "type": "page",
        },
        "index": {
          "hidden": true,
          "title": "Introduction",
          "type": "page",
        },
        "nextra_link": {
          "href": "https://github.com/shuding/nextra",
          "newWindow": true,
          "title": "Nextra ‚Üó",
          "type": "page",
        },
      },
      "name": "_meta.json",
    },
    {
      "children": [
        {
          "locale": "en-US",
          "meta": {
            "a-page": {
              "hidden": true,
              "theme": {
                "layout": "raw",
              },
              "title": "A Page",
              "type": "page",
            },
            "acknowledgement": "üß© Acknowledgement",
            "team": "üë• Team",
          },
          "name": "_meta.json",
        },
        {
          "locale": "en-US",
          "name": "a-page",
          "route": "/about/a-page",
        },
        {
          "locale": "en-US",
          "name": "acknowledgement",
          "route": "/about/acknowledgement",
        },
        {
          "locale": "en-US",
          "name": "team",
          "route": "/about/team",
        },
      ],
      "name": "about",
      "route": "/about",
    },
    {
      "children": [
        {
          "locale": "en-US",
          "meta": {
            "swr-v1": {
              "theme": {
                "breadcrumb": true,
              },
              "title": "Announcing SWR 1.0",
            },
          },
          "name": "_meta.json",
        },
        {
          "frontMatter": {
            "description": "Almost 2 years ago we open sourced SWR, the tiny data-fetching React library that people love. Today we are reaching another milestone: the 1.0 version of SWR.",
            "image": "https://assets.vercel.com/image/upload/v1630059453/swr/v1.png",
          },
          "locale": "en-US",
          "name": "swr-v1",
          "route": "/blog/swr-v1",
        },
      ],
      "name": "blog",
      "route": "/blog",
    },
    {
      "frontMatter": {
        "searchable": false,
      },
      "locale": "en-US",
      "name": "blog",
      "route": "/blog",
    },
    {
      "children": [
        {
          "locale": "en-US",
          "name": "404-500",
          "route": "/docs/404-500",
        },
        {
          "locale": "en-US",
          "meta": {
            "--- hey": {
              "title": "Getting Started",
              "type": "separator",
            },
            "--- my_new_separator": {
              "title": "Advanced",
              "type": "separator",
            },
            "404-500": "404/500 Custom Error Pages",
            "advanced": {
              "title": "Advanced",
            },
            "arguments": "Arguments",
            "change-log": {
              "theme": {
                "sidebar": false,
              },
              "title": "Change Log",
            },
            "conditional-fetching": "Conditional Data Fetching",
            "data-fetching": "Data Fetching",
            "error-handling": {
              "hidden": true,
              "title": "Error Handling",
            },
            "getting-started": "Getting Started",
            "github_link": {
              "href": "https://github.com/shuding/nextra",
              "newWindow": true,
              "title": "GitHub üêô ‚Üí",
              "type": "page",
            },
            "global-configuration": "Global Configuration",
            "middleware": "Middleware",
            "mutation": "Mutation",
            "options": "Options",
            "pagination": "Pagination",
            "prefetching": "Prefetching",
            "raw-layout": {
              "theme": {
                "layout": "raw",
              },
              "title": "Raw Layout",
            },
            "revalidation": "Auto Revalidation",
            "scroll-with-dynamic-height": "Scroll With Dynamic Height",
            "suspense": "Suspense",
            "typescript": "TypeScript :)",
            "with-nextjs": "Next.js SSG and SSR",
          },
          "name": "_meta.json",
        },
        {
          "children": [
            {
              "locale": "en-US",
              "meta": {
                "*": {
                  "theme": {
                    "footer": false,
                  },
                },
                "--- yoo": {
                  "title": "Do Not Use",
                  "type": "separator",
                },
                "cache": "Cache",
                "code-highlighting": "Code highlighting",
                "file-name.with.DOTS": "Filenames with dots",
                "markdown-import": "Markdown import",
                "more": "More: A Super Super Super Super Long Directory",
                "performance": "Performance",
                "react-native": "React Native",
              },
              "name": "_meta.json",
            },
            {
              "locale": "en-US",
              "name": "cache",
              "route": "/docs/advanced/cache",
            },
            {
              "locale": "en-US",
              "name": "code-highlighting",
              "route": "/docs/advanced/code-highlighting",
            },
            {
              "locale": "en-US",
              "name": "file-name.with.DOTS",
              "route": "/docs/advanced/file-name.with.DOTS",
            },
            {
              "locale": "en-US",
              "name": "markdown-import",
              "route": "/docs/advanced/markdown-import",
            },
            {
              "children": [
                {
                  "locale": "en-US",
                  "name": "loooooooooooooooooooong-title",
                  "route": "/docs/advanced/more/loooooooooooooooooooong-title",
                },
              ],
              "name": "more",
              "route": "/docs/advanced/more",
            },
            {
              "locale": "en-US",
              "name": "performance",
              "route": "/docs/advanced/performance",
            },
            {
              "locale": "en-US",
              "name": "react-native",
              "route": "/docs/advanced/react-native",
            },
          ],
          "name": "advanced",
          "route": "/docs/advanced",
        },
        {
          "locale": "en-US",
          "name": "advanced",
          "route": "/docs/advanced",
        },
        {
          "locale": "en-US",
          "name": "arguments",
          "route": "/docs/arguments",
        },
        {
          "locale": "en-US",
          "name": "change-log",
          "route": "/docs/change-log",
        },
        {
          "locale": "en-US",
          "name": "conditional-fetching",
          "route": "/docs/conditional-fetching",
        },
        {
          "locale": "en-US",
          "name": "data-fetching",
          "route": "/docs/data-fetching",
        },
        {
          "locale": "en-US",
          "name": "error-handling",
          "route": "/docs/error-handling",
        },
        {
          "locale": "en-US",
          "name": "getting-started",
          "route": "/docs/getting-started",
        },
        {
          "locale": "en-US",
          "name": "global-configuration",
          "route": "/docs/global-configuration",
        },
        {
          "locale": "en-US",
          "name": "middleware",
          "route": "/docs/middleware",
        },
        {
          "locale": "en-US",
          "name": "mutation",
          "route": "/docs/mutation",
        },
        {
          "locale": "en-US",
          "name": "options",
          "route": "/docs/options",
        },
        {
          "locale": "en-US",
          "name": "pagination",
          "route": "/docs/pagination",
        },
        {
          "locale": "en-US",
          "name": "prefetching",
          "route": "/docs/prefetching",
        },
        {
          "locale": "en-US",
          "name": "raw-layout",
          "route": "/docs/raw-layout",
        },
        {
          "locale": "en-US",
          "name": "revalidation",
          "route": "/docs/revalidation",
        },
        {
          "locale": "en-US",
          "name": "scroll-with-dynamic-height",
          "route": "/docs/scroll-with-dynamic-height",
        },
        {
          "locale": "en-US",
          "name": "suspense",
          "route": "/docs/suspense",
        },
        {
          "locale": "en-US",
          "name": "typescript",
          "route": "/docs/typescript",
        },
        {
          "locale": "en-US",
          "name": "with-nextjs",
          "route": "/docs/with-nextjs",
        },
      ],
      "name": "docs",
      "route": "/docs",
    },
    {
      "children": [
        {
          "locale": "en-US",
          "meta": {
            "auth": "Authentication",
            "basic": "Basic Usage",
            "error-handling": "Error Handling",
            "full": {
              "theme": {
                "layout": "raw",
              },
              "title": "Full",
            },
            "infinite-loading": "Infinite Loading",
            "ssr": "Next.js SSR",
          },
          "name": "_meta.json",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Authentication",
          },
          "locale": "en-US",
          "name": "auth",
          "route": "/examples/auth",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Basic Usage",
          },
          "locale": "en-US",
          "name": "basic",
          "route": "/examples/basic",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Error Handling",
          },
          "locale": "en-US",
          "name": "error-handling",
          "route": "/examples/error-handling",
        },
        {
          "locale": "en-US",
          "name": "full",
          "route": "/examples/full",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Infinite Loading",
          },
          "locale": "en-US",
          "name": "infinite-loading",
          "route": "/examples/infinite-loading",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Next.js SSR",
          },
          "locale": "en-US",
          "name": "ssr",
          "route": "/examples/ssr",
        },
      ],
      "name": "examples",
      "route": "/examples",
    },
    {
      "frontMatter": {
        "searchable": false,
        "title": "React Hooks for Data Fetching",
      },
      "locale": "en-US",
      "name": "index",
      "route": "/",
    },
  ],
  "/docs/data-fetching",
  "Data Fetching",
]
`;

exports[`Page Process > pageMap zh-CN 1`] = `
[
  [
    {
      "locale": "zh-CN",
      "meta": {
        "blog": {
          "title": "ÂçöÂÆ¢",
          "type": "page",
        },
        "docs": {
          "title": "ÊñáÊ°£",
          "type": "page",
        },
        "examples": {
          "title": "Á§∫‰æã",
          "type": "page",
        },
        "index": {
          "hidden": true,
          "title": "ÁÆÄ‰ªã",
          "type": "page",
        },
      },
      "name": "_meta.json",
    },
    {
      "children": [
        {
          "locale": "en-US",
          "meta": {
            "a-page": {
              "hidden": true,
              "theme": {
                "layout": "raw",
              },
              "title": "A Page",
              "type": "page",
            },
            "acknowledgement": "üß© Acknowledgement",
            "team": "üë• Team",
          },
          "name": "_meta.json",
        },
        {
          "locale": "en-US",
          "name": "a-page",
          "route": "/about/a-page",
        },
        {
          "locale": "en-US",
          "name": "acknowledgement",
          "route": "/about/acknowledgement",
        },
        {
          "locale": "en-US",
          "name": "team",
          "route": "/about/team",
        },
      ],
      "name": "about",
      "route": "/about",
    },
    {
      "children": [
        {
          "locale": "zh-CN",
          "meta": {
            "swr-v1": "SWR 1.0 ÂèëÂ∏É",
          },
          "name": "_meta.json",
        },
        {
          "frontMatter": {
            "description": "Â§ßÁ∫¶‰∏§Âπ¥ÂâçÔºåÊàë‰ª¨ÂºÄÊ∫ê‰∫Ü SWRÔºåÂ§ßÂÆ∂ÂñúÁà±ÁöÑÂ∞èÂûãÊï∞ÊçÆËØ∑Ê±Ç React Â∫ì„ÄÇ‰ªäÂ§©ÔºåÊàë‰ª¨ËøéÊù•‰∫ÜÂè¶‰∏Ä‰∏™ÈáåÁ®ãÁ¢ëÔºöSWR 1.0 ÁâàÊú¨ÂèëÂ∏É„ÄÇ",
            "image": "https://assets.vercel.com/image/upload/v1630059453/swr/v1.png",
          },
          "locale": "zh-CN",
          "name": "swr-v1",
          "route": "/blog/swr-v1",
        },
      ],
      "name": "blog",
      "route": "/blog",
    },
    {
      "children": [
        {
          "locale": "zh-CN",
          "meta": {
            "advanced": "È´òÁ∫ß",
            "arguments": "‰º†ÂÖ•ÂèÇÊï∞",
            "change-log": "Êõ¥Êñ∞Êó•Âøó",
            "conditional-fetching": "Êù°‰ª∂Êï∞ÊçÆËØ∑Ê±Ç",
            "data-fetching": "Êï∞ÊçÆËØ∑Ê±Ç",
            "error-handling": "ÈîôËØØÂ§ÑÁêÜ",
            "getting-started": "ÂÖ•Èó®",
            "global-configuration": "ÂÖ®Â±ÄÈÖçÁΩÆ",
            "middleware": "‰∏≠Èó¥‰ª∂",
            "mutation": "Êï∞ÊçÆÊõ¥Êîπ",
            "options": "ÈÄâÈ°π",
            "pagination": "ÂàÜÈ°µ",
            "prefetching": "È¢ÑËØ∑Ê±Ç",
            "revalidation": "Ëá™Âä®ÈáçÊñ∞ËØ∑Ê±Ç",
            "scroll-with-dynamic-height": "‰ª•ÂãïÊÖãÈ´òÂ∫¶ÊªæÂãï",
            "suspense": "Suspense",
            "typescript": "Typescript",
            "with-nextjs": "Next.js SSG Âíå SSR",
          },
          "name": "_meta.json",
        },
        {
          "children": [
            {
              "locale": "zh-CN",
              "meta": {
                "cache": "ÁºìÂ≠ò",
                "file-name.with.DOTS": "Â∏¶ÁÇπÁöÑÊñá‰ª∂Âêç",
                "performance": "ÊÄßËÉΩ",
                "react-native": "React Native",
              },
              "name": "_meta.json",
            },
            {
              "locale": "zh-CN",
              "name": "cache",
              "route": "/docs/advanced/cache",
            },
            {
              "locale": "zh-CN",
              "name": "file-name.with.DOTS",
              "route": "/docs/advanced/file-name.with.DOTS",
            },
            {
              "children": [
                {
                  "locale": "en-US",
                  "name": "loooooooooooooooooooong-title",
                  "route": "/docs/advanced/more/loooooooooooooooooooong-title",
                },
              ],
              "name": "more",
              "route": "/docs/advanced/more",
            },
            {
              "locale": "zh-CN",
              "name": "performance",
              "route": "/docs/advanced/performance",
            },
            {
              "locale": "zh-CN",
              "name": "react-native",
              "route": "/docs/advanced/react-native",
            },
            {
              "locale": "en-US",
              "name": "code-highlighting",
              "route": "/docs/advanced/code-highlighting",
            },
            {
              "locale": "en-US",
              "name": "markdown-import",
              "route": "/docs/advanced/markdown-import",
            },
          ],
          "name": "advanced",
          "route": "/docs/advanced",
        },
        {
          "locale": "zh-CN",
          "name": "arguments",
          "route": "/docs/arguments",
        },
        {
          "locale": "zh-CN",
          "name": "change-log",
          "route": "/docs/change-log",
        },
        {
          "locale": "zh-CN",
          "name": "conditional-fetching",
          "route": "/docs/conditional-fetching",
        },
        {
          "locale": "zh-CN",
          "name": "data-fetching",
          "route": "/docs/data-fetching",
        },
        {
          "locale": "zh-CN",
          "name": "error-handling",
          "route": "/docs/error-handling",
        },
        {
          "locale": "zh-CN",
          "name": "getting-started",
          "route": "/docs/getting-started",
        },
        {
          "locale": "zh-CN",
          "name": "global-configuration",
          "route": "/docs/global-configuration",
        },
        {
          "locale": "zh-CN",
          "name": "middleware",
          "route": "/docs/middleware",
        },
        {
          "locale": "zh-CN",
          "name": "mutation",
          "route": "/docs/mutation",
        },
        {
          "locale": "zh-CN",
          "name": "options",
          "route": "/docs/options",
        },
        {
          "locale": "zh-CN",
          "name": "pagination",
          "route": "/docs/pagination",
        },
        {
          "locale": "zh-CN",
          "name": "prefetching",
          "route": "/docs/prefetching",
        },
        {
          "locale": "zh-CN",
          "name": "revalidation",
          "route": "/docs/revalidation",
        },
        {
          "locale": "zh-CN",
          "name": "scroll-with-dynamic-height",
          "route": "/docs/scroll-with-dynamic-height",
        },
        {
          "locale": "zh-CN",
          "name": "suspense",
          "route": "/docs/suspense",
        },
        {
          "locale": "zh-CN",
          "name": "with-nextjs",
          "route": "/docs/with-nextjs",
        },
        {
          "locale": "en-US",
          "name": "404-500",
          "route": "/docs/404-500",
        },
        {
          "locale": "en-US",
          "name": "advanced",
          "route": "/docs/advanced",
        },
        {
          "locale": "en-US",
          "name": "raw-layout",
          "route": "/docs/raw-layout",
        },
        {
          "locale": "en-US",
          "name": "typescript",
          "route": "/docs/typescript",
        },
      ],
      "name": "docs",
      "route": "/docs",
    },
    {
      "children": [
        {
          "locale": "zh-CN",
          "meta": {
            "auth": "Ë∫´‰ªΩÈ™åËØÅ",
            "basic": "Âü∫Êú¨Áî®Ê≥ï",
            "error-handling": "ÈîôËØØÂ§ÑÁêÜ",
            "infinite-loading": "Êó†ÈôêÂä†ËΩΩ",
            "ssr": "Next.js SSR",
          },
          "name": "_meta.json",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Ë∫´‰ªΩÈ™åËØÅ",
          },
          "locale": "zh-CN",
          "name": "auth",
          "route": "/examples/auth",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Âü∫Êú¨Áî®Ê≥ï",
          },
          "locale": "zh-CN",
          "name": "basic",
          "route": "/examples/basic",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "ÈîôËØØÂ§ÑÁêÜ",
          },
          "locale": "zh-CN",
          "name": "error-handling",
          "route": "/examples/error-handling",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Êó†ÈôêÂä†ËΩΩ",
          },
          "locale": "zh-CN",
          "name": "infinite-loading",
          "route": "/examples/infinite-loading",
        },
        {
          "locale": "en-US",
          "name": "full",
          "route": "/examples/full",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Next.js SSR",
          },
          "locale": "en-US",
          "name": "ssr",
          "route": "/examples/ssr",
        },
      ],
      "name": "examples",
      "route": "/examples",
    },
    {
      "frontMatter": {
        "title": "Áî®‰∫éÊï∞ÊçÆËØ∑Ê±ÇÁöÑ React Hooks Â∫ì",
      },
      "locale": "zh-CN",
      "name": "index",
      "route": "/",
    },
    {
      "locale": "",
      "name": "404",
      "route": "/404",
    },
    {
      "locale": "",
      "name": "500",
      "route": "/500",
    },
    {
      "frontMatter": {
        "searchable": false,
      },
      "locale": "en-US",
      "name": "blog",
      "route": "/blog",
    },
  ],
  "/docs/data-fetching",
  "Êï∞ÊçÆËØ∑Ê±Ç",
]
`;
